src/git.rs

```diff
@@ -79,40 +115,181 @@ impl Git for LiveGit {
+
+        for commit in commits.iter() {
+            let author = commit.author();
+            let commit_email = match author.email() {
+                Some(e) => e,
+                None => return Err(Box::new(CapnError::new(format!("Email on commit not found"))))
+            };
+            let fingerprint = team_fingerprints.iter().find(|f| f.email == commit_email);
+            match fingerprint{
+                Some(f) => commit_fingerprints.insert(f.id.to_string()),
+                None => return Err(Box::new(CapnError::new(format!("Team fingerprint not found"))))
+            };
+        }
+        Ok(commit_fingerprints)
+    }
+
+    fn pushed(&self, commit_id: Oid) -> Result<bool, Box<dyn Error>> {
+        debug!("Check if commit {} has already been pushed", commit_id);
+
+        let repo_path = self.repo.path();
+        debug!("Repo path {:?}", repo_path);
+        let result = Command::new("git")
+            .current_dir(repo_path)
+            .arg("branch")
+            .arg("--contains")
+            .arg(commit_id.to_string())
+            .output()?;
+        debug!("RESULT {:?}", result);
+        if !result.status.success() {
+            return Err(Box::new(CapnError::new(format!("Call to git branch contains failed for commit {} with status {}", commit_id,result.status))));
+        }
+
+        let output = String::from_utf8(result.stdout)?;
+        match output.trim() {
+            "" => return Ok(false),
+            _ => return Ok(true)
+        };                
+    }
+
+    fn find_commits(&self, from_id: Oid, to_id: Oid) -> Result<Vec<Commit<'_>>, Box<dyn Error>> {
+        debug!("Find commits between {} to {}", from_id, to_id);
+
+        let mut v = Vec::new();
+        let new_commit = self.repo.find_commit(to_id)?;
+        let mut current_id = to_id;
+        v.push(new_commit);
+        while current_id != from_id {
+            let current_commit = self.repo.find_commit(current_id)?;
+            for parent in current_commit.parents() {
+                current_id = parent.id();
+                let parent_commit = self.repo.find_commit(parent.id())?;
+                if current_id != from_id  {
+                    v.push(parent_commit);
+                }
+            }      
+        }
+        debug!("Commits found {:#?}",v);
+        Ok(v)         
+    }
+
+    fn find_unpushed_commits(&self, new_commit_id: Oid) -> Result<Vec<Commit<'_>>, Box<dyn Error>> {
+        debug!("Get unpushed commits from {} ", new_commit_id);
+
+        let mut v = Vec::new();
+        let new_commit = self.repo.find_commit(new_commit_id)?;
+        let mut current_id = new_commit_id;
+        let mut parent_count = new_commit.parent_count();
+        let mut pushed = self.pushed(current_id)?;
+        if !pushed {
+            v.push(new_commit);
         }
-        fn read_file(&self, _path: &str) -> Result<String, Box<dyn Error>> {
-            Ok(String::from(""))
+        while !pushed && parent_count > 0 {
```

Comment:
This loop looks very similar to the previous one.  Can't we create a commit iterator which would contain the logic of iterating over a linear history of commits and fails at merge commits.
You could then use collections operators to greatly simplify this code.